R/epi_spread.R

```diff
@@ -0,0 +1,53 @@
+#' @title Prepare contact and demography data for multiple risk groups
+#'
+#' @param contact_matrix Social contact matrix. Entry $mm_{ij}$ gives average
+#' number of contacts in group $i$ reported by participants in group j
+#' @param demography_vector Demography vector. Entry $pp_{i}$ gives proportion
+#' of total population in group $i$ (model will normalise if needed)
+#' @param p_susceptibility A matrix giving the probability that an individual
+#' in demography group $i$ is in risk (or susceptibility) group $j$.
+#' Each row represents the overall distribution of individuals in demographic
+#' group $i$ across risk groups, and each row *must sum to 1.0*.
+#' @param susceptibility A matrix giving the susceptibility of individuals in
+#' demographic group $i$ and risk group $j$.
+#'
+#' @return A list object with named elements: "contact_matrix",
+#' "demography_vector", "p_susceptibility_", and "susceptibility".
+#' The contact matrix is replicated row and column wise for each risk group
+#' and the demography vector is replicated for each risk group.
+epi_spread <- function(contact_matrix,
+                       demography_vector,
+                       p_susceptibility,
+                       susceptibility) {
+  # count risk groups
+  n_susc_groups <- ncol(p_susceptibility)
+  # make p_susceptibility matrix of ones
+  p_susceptibility_ <- matrix(
+    1.0,
+    nrow = prod(dim(p_susceptibility)), ncol = 1
+  )
+  # make lps, a 1 col matrix of all p_susc values
+  lps <- matrix(p_susceptibility, nrow = prod(dim(p_susceptibility)), ncol = 1)
+  # replicate the demography vector and multiply by p_susceptibility
+  demography_vector_spread <- rep(demography_vector, n_susc_groups)
+  demography_vector_spread <- demography_vector_spread * as.vector(lps)
+
+  # replicate contact matrix
```

Comment:
see `?kronecker` function. Not sure if that would be a little quicker or if it just wraps something similar. May or may not be clearer using %x% / kronecker.

Reply:
The Kronecker product would give an array with the dimensions `length(demography_vector) * length(p_susceptibility)` (the second term is equivalent to `length(lps)`), but we need the dimensions to be `length(demography_vector) * n_susc_groups`.

Keeping the current syntax with minor changes from converting `lps` to vector earlier. Please comment if I'm wrong here!

Reply:
I'm talking about the `rbind` bits (confusing on GitHub comments)
```r
  # replicate contact matrix
  contact_matrix_ <- kronecker(
    X = matrix(1, nrow = n_susc_groups, ncol = n_susc_groups),
    Y = contact_matrix
  )
```

Reply:
^ this may not actually be quicker - just felt a little cleaner .... maybe ... no strong feelings either way.

Reply:
It does look cleaner, and I don't imagine it would be much slower - could you make an edit suggestion so that I can accept it as a commit - you would then be correctly attributed as author on the PR? No worries if too much trouble!

Reply:
Issue opened as #49 and fixed